### PR TITLE
Kernel function Test Cases

### PR DESCRIPTION
**Pull request by SantamRC**
To run ,
`python dev/generate-strategies.py`
`python hypothesis-tests-spec/gen.py`
`python -m pytest  hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py`

### CODE REVIEW

File: dev/generate-strategies.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+from hypothesis.strategies import composite
+import os
+import json
+from shutil import copy
+
+
+CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))
+
+
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
```
Comment (ianna) [inline dev/generate-strategies.py]: this should be the first line in a file

**Review by ianna — COMMENTED** [review]
@SantamRC - great! The same number of tests run 50% faster - if I add the number of these cases to 25 with the test data from the current test `tests-cpu-kernels/test_cpuawkward_IndexedArray64_validity.py`:
```python
% python -m pytest hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py
==================================================================================================== test session starts =====================================================================================================
platform darwin -- Python 3.8.3, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: awkward-1.0, inifile: setup.cfg
plugins: hypothesis-6.13.0, anyio-3.1.0
collected 1 item                                                                                                                                                                                                             

hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py .                                                                                                                                                         [100%]

===================================================================================================== 1 passed in 0.29s ======================================================================================================
```
the above output is misleading though - this one tests is testing 25 cases.

```python
% python -m pytest tests-cpu-kernels/test_cpuawkward_IndexedArray64_validity.py
==================================================================================================== test session starts =====================================================================================================
platform darwin -- Python 3.8.3, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: awkward-1.0, inifile: setup.cfg
plugins: hypothesis-6.13.0, anyio-3.1.0
collected 25 items                                                                                                                                                                                                           

tests-cpu-kernels/test_cpuawkward_IndexedArray64_validity.py .........................                                                                                                                                 [100%]

===================================================================================================== 25 passed in 0.73s =====================================================================================================

```

Some general comments about the approach:

I think, we need to follow closely the current workflow approach. For example, the same test run from `tests-cpu-kernels/test_cpuawkward_IndexedArray64_validity.py` clearly shows what are the input values for each of the 25 tests it would:
```python
def test_cpuawkward_IndexedArray64_validity_2():
    index = [1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1]
    index = (ctypes.c_int64*len(index))(*index)
    length = 3
    lencontent = 2
    isoption = False
    funcC = getattr(lib, 'awkward_IndexedArray64_validity')
    ret_pass = funcC(index, length, lencontent, isoption)
    assert not ret_pass.str
```
The advantage of this is that it is clear which test values cause an error. For example, if I modify it to something non-valid:
```python
def test_cpuawkward_IndexedArray64_validity_1():
    index = [-1000, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1]
    index = (ctypes.c_int64*len(index))(*index)
    length = 100000
    lencontent = -12
    isoption = True
    funcC = getattr(lib, 'awkward_IndexedArray64_validity')
    ret_pass = funcC(index, length, lencontent, isoption)
    assert not ret_pass.str
```
It would tell me explicitly:
```python
tests-cpu-kernels/test_cpuawkward_IndexedArray64_validity.py F........................                                                                                                                                 [100%]

========================================================================================================== FAILURES ==========================================================================================================
_________________________________________________________________________________________ test_cpuawkward_IndexedArray64_validity_1 __________________________________________________________________________________________

    def test_cpuawkward_IndexedArray64_validity_1():
        index = [-1000, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1]
        index = (ctypes.c_int64*len(index))(*index)
        length = 100000
        lencontent = -12
        isoption = True
        funcC = getattr(lib, 'awkward_IndexedArray64_validity')
        ret_pass = funcC(index, length, lencontent, isoption)
>       assert not ret_pass.str
E       AssertionError: assert not b'index[i] >= len(content)'
E        +  where b'index[i] >= len(content)' = <awkward._kernel_signatures.ERROR object at 0x7fa6389e87c0>.str

tests-cpu-kernels/test_cpuawkward_IndexedArray64_validity.py:24: AssertionError
================================================================================================== short test summary info ===================================================================================================
FAILED tests-cpu-kernels/test_cpuawkward_IndexedArray64_validity.py::test_cpuawkward_IndexedArray64_validity_1 - AssertionError: assert not b'index[i] >= len(content)'
================================================================================================ 1 failed, 24 passed in 0.73s ================================================================================================
```
BTW, you need to remove the last call - `pytest` will do it for you and the errors would be reported better:
```python
========================================================================================================== FAILURES ==========================================================================================================
_____________________________________________________________________________________________ test_awkward_IndexedArray_validity _____________________________________________________________________________________________

    def test_awkward_IndexedArray_validity():
        f = open(
            os.path.join(CURRENT_DIR, "..", "test-data.json"),
        )
        data = json.load(f)
        for d in data["datas"]:
            x = d["input"]
            y = d["output"]
>           result = kernels.awkward_IndexedArray_validity(*tuple(x))

hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py:16: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

index = [0], length = 10, lencontent = 1, isoption = False

    def awkward_IndexedArray_validity(index, length, lencontent, isoption):
        for i in range(length):
>           idx = index[i]
E           IndexError: list index out of range

hypothesis-tests-spec/kernels.py:737: IndexError
================================================================================================== short test summary info ===================================================================================================
FAILED hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py::test_awkward_IndexedArray_validity - IndexError: list index out of range
===================================================================================================== 1 failed in 0.37s ======================================================================================================
```
instead of
```python
=========================================================================================================== ERRORS ===========================================================================================================
_______________________________________________________________________ ERROR collecting hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py ________________________________________________________________________
hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py:20: in <module>
    test_awkward_IndexedArray_validity()
hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py:16: in test_awkward_IndexedArray_validity
    result = kernels.awkward_IndexedArray_validity(*tuple(x))
hypothesis-tests-spec/kernels.py:737: in awkward_IndexedArray_validity
    idx = index[i]
E   IndexError: list index out of range
================================================================================================== short test summary info ===================================================================================================
ERROR hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py - IndexError: list index out of range
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
====================================================================================================== 1 error in 0.32s ======================================================================================================
```


**Conclusion**
Proposed approach gives us a possibility to run twice as many tests without increasing testing time. This is great! However, we should discuss how it scales with the number of roles since not all tests would have the same number of inputs. Perhaps, the json test data fragment relevant to the test should be incorporated in the test itself. The generated test could look like:
```python
import kernels

def test_awkward_IndexedArray_validity_1():
        input = [
                [0],
                10,
                1,
                False
        ]
        output = None
        result = kernels.awkward_IndexedArray_validity(*tuple(input))
        assert result == output
```
then we'd see 
```python
% python -m pytest hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py
==================================================================================================== test session starts =====================================================================================================
platform darwin -- Python 3.8.3, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: awkward-1.0, inifile: setup.cfg
plugins: hypothesis-6.13.0, anyio-3.1.0
collected 1 item                                                                                                                                                                                                             

hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py F                                                                                                                                                         [100%]

========================================================================================================== FAILURES ==========================================================================================================
____________________________________________________________________________________________ test_awkward_IndexedArray_validity_1 ____________________________________________________________________________________________

    def test_awkward_IndexedArray_validity_1():
    	input = [
        		[0],
        		10,
        		1,
        		False
       	]
    	output = 0
>   	result = kernels.awkward_IndexedArray_validity(*tuple(input))

hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py:11: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

index = [0], length = 10, lencontent = 1, isoption = False

    def awkward_IndexedArray_validity(index, length, lencontent, isoption):
        for i in range(length):
>           idx = index[i]
E           IndexError: list index out of range

hypothesis-tests-spec/kernels.py:737: IndexError
================================================================================================== short test summary info ===================================================================================================
FAILED hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py::test_awkward_IndexedArray_validity_1 - IndexError: list index out of range
===================================================================================================== 1 failed in 0.35s ======================================================================================================
```
**Next steps**

- [ ] Generate more realistic data samples as the ones I see in json look the same and do not test much:

```json
   {
   "input": [
    [
     0
    ],
    1,
    1,
    false
   ],
   "output": null
  },
```

File: hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    )
+    data = json.load(f)
+    for d in data["datas"]:
+        x = d["input"]
+        y = d["output"]
+        result = kernels.awkward_IndexedArray_validity(*tuple(x))
+        assert result == y
+
+
+test_awkward_IndexedArray_validity()
```
Comment (ianna) [inline hypothesis-tests-spec/tests_awkward_IndexedArray_validity.py]: @SantamRC - this call is not needed, `pytest` will take care of it as long as the test is named with `test_` prefix.

**Review by ianna — COMMENTED** [review]
a general comment: please, use `pre-commit run --all` before pushing. Thanks

**Comment by SantamRC** [comment]
@ianna just running `python dev/generate-strategies.py` and then running `pytest hypothesis-tests-spec` will run it. I am working on producing more better values and adding other IndexedArray kernel functions to this script

**Comment by SantamRC** [comment]
@ianna please have a look at the input data in the json file (the error cases and the normal ones)

**Comment by SantamRC** [comment]
Continuing work in #1044

**Pull request closed without merging**